File: src/main/java/com/github/tianchenghang/tools/FileOperationTool.java
package com.github.tianchenghang.tools;

import cn.hutool.core.io.FileUtil;
import com.github.tianchenghang.constants.FileConstant;
import org.springframework.ai.tool.annotation.Tool;
import org.springframework.ai.tool.annotation.ToolParam;

public class FileOperationTool {

  private final String FILE_OUTPUT_DIR = FileConstant.FILE_OUTPUT_DIR + "/file";

  @Tool(description = "Read content from a file")
  public String readFile(@ToolParam(description = "Name of the file to read") String filename) {
    var filepath = FILE_OUTPUT_DIR + "/" + filename;
    try {
      return FileUtil.readUtf8String(filepath);
    } catch (Exception e) {
      return "Error reading file: " + e.getMessage();
    }
  }

  @Tool(description = "Write content to a file")
  public String writeFile(
      @ToolParam(description = "Name of the file to write") String filename,
      @ToolParam(description = "Content to write to the file") String content) {
    var filepath = FILE_OUTPUT_DIR + "/" + filename;

    try {
      FileUtil.mkdir(FILE_OUTPUT_DIR);
      FileUtil.writeUtf8String(content, filepath);
      return "File written successfully to: " + filepath;
    } catch (Exception e) {
      return "Error writing to file: " + e.getMessage();
    }
  }
}

File: server/src/tools/file-operation-tool.ts
import { join } from 'path';
import { StructuredTool } from '@langchain/core/tools';
import { z } from 'zod';
import { readFile, writeFile } from 'fs/promises';
import ensureDir from './ensure-dir';

const OUTPUT_DIR = join(process.cwd(), './tmp/file');

export class ReadFileTool extends StructuredTool {
  name = ReadFileTool.name;
  description = 'Read content from a file';
  schema = z.object({
    filename: z.string().describe('Name of the file to read'),
  });

  async _call(args: { filename: string }): Promise<string> {
    const { filename } = args;
    const filepath = join(OUTPUT_DIR, filename);
    try {
      const content = await readFile(filepath, 'utf-8');
      return content;
    } catch (err) {
      console.error('Reading file error:', err);
      return 'Reading file error';
    }
  }
}

export class WriteFileTool extends StructuredTool {
  name = 'WriteFileTool';
  description = 'Write content to a file';
  schema = z.object({
    filename: z.string().describe('Name of the file to write'),
    content: z.string().describe('Content to write to the file'),
  });

  async _call(args: { filename: string; content: string }): Promise<string> {
    const { filename, content } = args;
    const filepath = join(OUTPUT_DIR, filename);
    try {
      await ensureDir(OUTPUT_DIR);
      await writeFile(filepath, content, 'utf-8');
      return `File written successfully to: ${filepath}`;
    } catch (err) {
      console.error('Writing file error:', err);
      return 'Writing file error';
    }
  }
}
